testing/checkpoints/test_checkpoint_pipeline_runner.py

```diff
@@ -0,0 +1,134 @@
+import numpy as np
+import pytest
+
+from neuraxle.checkpoints.base_checkpoint_step import BaseCheckpointStep
+from neuraxle.checkpoints.checkpoint_pipeline_runner import CheckpointPipelineRunner
+from neuraxle.pipeline import Pipeline
+from neuraxle.steps.util import TransformCallbackStep, TapeCallbackFunction
+
+
+class SomeCheckpointStep(BaseCheckpointStep):
+    def __init__(self, checkpoint_data_inputs, checkpoint_expected_outputs):
+        super().__init__()
+        self.checkpoint_data_inputs = checkpoint_data_inputs
+        self.checkpoint_expected_outputs = checkpoint_expected_outputs
+        self.saved = False
+        self.data_inputs_checkpoint = None
+        self.expected_outputs_checkpoint = None
+
+    def load_checkpoint(self):
+        return self.checkpoint_data_inputs, self.checkpoint_expected_outputs
+
+    def save_checkpoint(self, data_inputs, expected_outputs=None):
+        self.data_inputs_checkpoint = data_inputs
+        self.expected_outputs_checkpoint = expected_outputs
+        self.saved = True
+
+
+data_inputs = np.ones((1, 1))
+expected_outputs = np.ones((1, 1))
+chekpoint = SomeCheckpointStep(data_inputs, expected_outputs)
+chekpoint_not_saved = SomeCheckpointStep(None, None)
+tape = TapeCallbackFunction()
+
+tape_without_checkpoint_test_arguments = (
+    [TransformCallbackStep(tape.callback, ["1"]),
+     TransformCallbackStep(tape.callback, ["2"]),
+     TransformCallbackStep(tape.callback, ["3"])],
+    ["1", "2", "3"])
+
+tape_checkpoint_not_saved_test_arguments = (
+    [TransformCallbackStep(tape.callback, ["1"]),
+     SomeCheckpointStep(None, None),
+     TransformCallbackStep(tape.callback, ["2"]),
+     TransformCallbackStep(tape.callback, ["3"])],
+    ["1", "2", "3"])
+
+tape_checkpoint_saved_after_first_step_test_arguments = (
+    [TransformCallbackStep(tape.callback, ["1"]),
+     SomeCheckpointStep(data_inputs, expected_outputs),
+     TransformCallbackStep(tape.callback, ["2"]),
+     TransformCallbackStep(tape.callback, ["3"])],
+    ["2", "3"])
+
+tape_checkpoint_saved_after_second_step_test_arguments = (
+    [TransformCallbackStep(tape.callback, ["1"]),
+     TransformCallbackStep(tape.callback, ["2"]),
+     SomeCheckpointStep(data_inputs, expected_outputs),
+     TransformCallbackStep(tape.callback, ["3"])],
+    ["3"])
```

Comment:
I'm not sure if my suggestion will be useful here, but wouldn't this ["3"] list here instead be ["1", "2", "3", "3"] ? 
(E.g.: it does 1, 2 and 3, then upon passing data a second time in the pipeline it does just 3, hence 1, 2, 3, 3). 

At least, the test with just 3 instead of 1,2,3,3 still seems good!

Reply:
the class SomeCheckpointSteps takes arguments to mock existing checkpoints. this is why I passed data_inputs, and expected_outputs in the constructor. I will name the constructor parameters to make it more clear i.e data_inputs_checkpoint=, expected_outputs_checkpoint=